// @dotenvx/dotenvx-ops/src/lib/api/postSync.js
const { http } = require('../../lib/helpers/http')
const buildApiError = require('../../lib/helpers/buildApiError')
const packageJson = require('../../lib/helpers/packageJson')

class PostSync {
  constructor (hostname, token, devicePublicKey, encoded, dotenvxProjectId, pwd = null, gitUrl = null, gitBranch = null, systemUuid = null, osPlatform = null, osArch = null, force = false) {
    this.hostname = hostname || 'https://ops.dotenvx.com'
    this.token = token
    this.devicePublicKey = devicePublicKey
    this.encoded = encoded
    this.dotenvxProjectId = dotenvxProjectId
    this.pwd = pwd
    this.gitUrl = gitUrl
    this.gitBranch = gitBranch
    this.systemUuid = systemUuid
    this.osPlatform = osPlatform
    this.osArch = osArch
    this.force = force
  }

  async run () {
    const token = this.token
    const devicePublicKey = this.devicePublicKey
    const url = `${this.hostname}/api/sync`
    const encoded = this.encoded
    const dotenvxProjectId = this.dotenvxProjectId
    const syncedAt = new Date().toISOString()
    const pwd = this.pwd
    const gitUrl = this.gitUrl
    const gitBranch = this.gitBranch
    const systemUuid = this.systemUuid
    const osPlatform = this.osPlatform
    const osArch = this.osArch
    const force = this.force

    const resp = await http(url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        device_public_key: devicePublicKey,
        encoded,
        dotenvx_project_id: dotenvxProjectId,
        synced_at: syncedAt,
        pwd,
        git_url: gitUrl,
        git_branch: gitBranch,
        system_uuid: systemUuid,
        os_platform: osPlatform,
        os_arch: osArch,
        cli_version: packageJson.version,
        force
      })
    })

    const json = await resp.body.json()

    if (resp.statusCode >= 400) {
      throw buildApiError(resp.statusCode, json)
    }

    return json
  }
}

module.exports = PostSync
