// @dotenvx/dotenvx-ops/src/lib/api/postRotateConnect.js
const { http } = require('../../lib/helpers/http')
const buildApiError = require('../../lib/helpers/buildApiError')

class PostRotateConnect {
  constructor (hostname, token, org, slug, username, password, email = null, playwrightStorageStateStringified) {
    this.hostname = hostname || 'https://ops.dotenvx.com'
    this.token = token

    this.org = org
    this.slug = slug
    this.username = username
    this.password = password
    this.email = email
    this.playwrightStorageStateStringified = playwrightStorageStateStringified
  }

  async run () {
    const token = this.token
    const url = `${this.hostname}/api/rotate/connect`

    const org = this.org
    const slug = this.slug
    const username = this.username
    const password = this.password
    const email = this.email
    const playwrightStorageStateStringified = this.playwrightStorageStateStringified

    const resp = await http(url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        org,
        slug,
        username,
        password,
        email,
        playwright_storage_state: playwrightStorageStateStringified
      })
    })

    const json = await resp.body.json()

    if (resp.statusCode >= 400) {
      throw buildApiError(resp.statusCode, json)
    }

    return json
  }
}

module.exports = PostRotateConnect
