// @dotenvx/dotenvx-ops/src/lib/api/postOauthDeviceCode.js
const { http } = require('../../lib/helpers/http')

// const Device = require('../../db/device')
const buildOauthError = require('../../lib/helpers/buildOauthError')

const OAUTH_CLIENT_ID = 'oac_dotenvxcli'

class PostOauthDeviceCode {
  constructor (hostname, devicePublicKey, systemInformation, dotenvxProjectId = null) {
    this.hostname = hostname
    this.devicePublicKey = devicePublicKey
    this.systemInformation = systemInformation
    this.dotenvxProjectId = dotenvxProjectId
  }

  async run () {
    const url = `${this.hostname}/oauth/device/code`
    const devicePublicKey = this.devicePublicKey
    const systemInformation = this.systemInformation
    const dotenvxProjectId = this.dotenvxProjectId

    const resp = await http(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        client_id: OAUTH_CLIENT_ID,
        device_public_key: devicePublicKey,
        system_information: systemInformation,
        dotenvx_project_id: dotenvxProjectId
      })
    })

    const json = await resp.body.json()

    if (resp.statusCode >= 400) {
      throw buildOauthError(resp.statusCode, json)
    }

    return json
  }
}

module.exports = PostOauthDeviceCode
