// @dotenvx/dotenvx-ops/src/lib/services/rotateNpmConnect.js
const PostRotateConnect = require('./../api/postRotateConnect')

const playwrightConnect = require('./../helpers/playwrightConnect')

const TIMEOUT = 500 // visibility timeout
const SLUG = 'npm'

class RotateNpmConnect {
  constructor (hostname, token, org, username, password, email = null) {
    this.hostname = hostname
    this.org = org
    this.token = token
    this.username = username
    this.password = password

    // optional
    this.email = email
  }

  async run () {
    const hostname = this.hostname
    const token = this.token
    const org = this.org
    const username = this.username
    const password = this.password
    const email = this.email

    const { browser, context, page } = await playwrightConnect()

    const usernameSelector = 'input[type="text"]'
    const passwordSelector = 'input[type="password"]'

    await page.goto('https://npmjs.com/login', { waitUntil: 'domcontentloaded' })
    await page.waitForTimeout(TIMEOUT)
    await page.fill(usernameSelector, username)
    await page.waitForTimeout(TIMEOUT)
    await page.fill(passwordSelector, password)
    await page.waitForTimeout(TIMEOUT)
    await page.press('input[type="password"]', 'Enter')
    await page.waitForNavigation({ timeout: 0 })
    await page.waitForSelector('img[alt="avatar"]', { state: 'visible', timeout: 0 })

    const playwrightStorageStateStringified = JSON.stringify(await context.storageState())

    const { uid, url } = await new PostRotateConnect(hostname, token, org, SLUG, username, password, email, playwrightStorageStateStringified).run()

    await browser.close()

    return {
      uid,
      url
    }
  }
}

module.exports = RotateNpmConnect
