// @dotenvx/dotenvx-ops/src/lib/services/sync.js
const fs = require('fs')
const path = require('path')
const si = require('systeminformation')
const dotenvx = require('@dotenvx/dotenvx')

const Session = require('./../../db/session')

const gitUrl = require('./../helpers/gitUrl')
const gitBranch = require('./../helpers/gitBranch')
const dotenvxProjectId = require('./../helpers/dotenvxProjectId')

// api calls
const PostSync = require('./../api/postSync')

class Sync {
  constructor (hostname, force = false) {
    this.hostname = hostname
    this.force = force
    this.cwd = process.cwd()
  }

  async run () {
    const sesh = new Session()
    const token = sesh.token()
    const devicePublicKey = sesh.devicePublicKey()

    // required
    // need to get all .env files here somehow
    const files = this._files()
    const payload = { files }
    const encoded = Buffer.from(JSON.stringify(payload)).toString('base64')
    const _dotenvxProjectId = dotenvxProjectId(this.cwd)

    // optional
    const _pwd = this.cwd
    const _gitUrl = gitUrl()
    const _gitBranch = gitBranch()

    const system = await si.system()
    const _systemUuid = system.uuid

    const osInfo = await si.osInfo()
    const _osPlatform = osInfo.platform
    const _osArch = osInfo.arch

    const data = await new PostSync(this.hostname, token, devicePublicKey, encoded, _dotenvxProjectId, _pwd, _gitUrl, _gitBranch, _systemUuid, _osPlatform, _osArch, this.force).run()

    return {
      id: data.id,
      dotenvxProjectId: data.dotenvx_project_id,
      projectUsernameName: data.project_username_name,
      files: data.files
    }
  }

  _files () {
    const out = []
    const filepaths = dotenvx.ls(this.cwd)

    for (const fp of filepaths) {
      const abs = path.join(this.cwd, fp)
      const src = fs.readFileSync(abs, 'utf8')
      out.push({ filepath: fp, src })
    }

    return out
  }
}

module.exports = Sync
