// @dotenvx/dotenvx-ops/src/db/session.js
const Conf = require('conf')
const dotenv = require('dotenv')
const si = require('systeminformation')

const Device = require('./device')
const jsonToEnv = require('./../lib/helpers/jsonToEnv')

class Session {
  constructor () {
    this.store = new Conf({
      cwd: process.env.DOTENVX_CONFIG || undefined,
      projectName: 'dotenvx',
      configName: '.env',
      projectSuffix: '',
      fileExtension: '',
      serialize: function (json) {
        return jsonToEnv(json)
      },
      // Convert .env format to an object
      deserialize: function (env) {
        return dotenv.parse(env)
      }
    })
  }

  status () {
    // if logged in
    if (this.username() && this.token()) {
      return 'on'
    }

    return 'off'
  }

  //
  // Get
  //
  hostname () {
    return this.store.get('DOTENVX_OPS_HOSTNAME') || this.store.get('DOTENVX_RADAR_HOSTNAME') || 'https://ops.dotenvx.com'
  }

  username () {
    return this.store.get('DOTENVX_OPS_USERNAME') || this.store.get('DOTENVX_RADAR_USERNAME') || undefined
  }

  token () {
    return this.store.get('DOTENVX_OPS_TOKEN') || this.store.get('DOTENVX_RADAR_TOKEN') || undefined
  }

  devicePublicKey () {
    return new Device().publicKey()
  }

  async systemInformation () {
    const system = await si.system()
    const osInfo = await si.osInfo()

    return {
      system_uuid: system.uuid,
      os_platform: osInfo.platform,
      os_arch: osInfo.arch
    }
  }

  //
  // Set/Delete
  //
  login (hostname, id, username, accessToken) {
    if (!hostname) {
      throw new Error('DOTENVX_OPS_HOSTNAME not set. Run [dotenvx-ops login]')
    }

    if (!id) {
      throw new Error('DOTENVX_OPS_USER not set. Run [dotenvx-ops login]')
    }

    if (!username) {
      throw new Error('DOTENVX_OPS_USERNAME not set. Run [dotenvx-ops login]')
    }

    if (!accessToken) {
      throw new Error('DOTENVX_OPS_TOKEN not set. Run [dotenvx-ops login]')
    }

    this.store.set('DOTENVX_OPS_USER', id)
    this.store.set('DOTENVX_OPS_USERNAME', username)
    this.store.set('DOTENVX_OPS_TOKEN', accessToken)
    this.store.set('DOTENVX_OPS_HOSTNAME', hostname)

    return accessToken
  }

  logout (hostname, id, accessToken) {
    if (!hostname) {
      throw new Error('DOTENVX_OPS_HOSTNAME not set. Run [dotenvx-ops login]')
    }

    if (!id) {
      throw new Error('DOTENVX_OPS_USER not set. Run [dotenvx-ops login]')
    }

    if (!accessToken) {
      throw new Error('DOTENVX_OPS_TOKEN not set. Run [dotenvx-ops login]')
    }

    this.store.delete('DOTENVX_OPS_USER')
    this.store.delete('DOTENVX_OPS_USERNAME')
    this.store.delete('DOTENVX_OPS_TOKEN')
    this.store.delete('DOTENVX_OPS_HOSTNAME')
    this.store.delete('DOTENVX_RADAR_USER')
    this.store.delete('DOTENVX_RADAR_USERNAME')
    this.store.delete('DOTENVX_RADAR_TOKEN')
    this.store.delete('DOTENVX_RADAR_HOSTNAME')

    return true
  }
}

module.exports = Session
